src/framework/edge.ml

```diff
@@ -0,0 +1,60 @@
+open Cil
+open Pretty
+
+type asm_out = (string option * string * CilType.Lval.t) list [@@deriving to_yojson]
+type asm_in  = (string option * string * CilType.Exp.t ) list [@@deriving to_yojson]
+
+type t =
+  | Assign of CilType.Lval.t * CilType.Exp.t
+  (** Assignments lval = exp *)
+  | Proc of CilType.Lval.t option * CilType.Exp.t * CilType.Exp.t list
+  (** Function calls of the form lva = fexp (e1, e2, ...) *)
+  | Entry of CilType.Fundec.t
+  (** Entry edge that relates function declaration to function body. You can use
+    * this to initialize the local variables. *)
+  | Ret of CilType.Exp.t option * CilType.Fundec.t
+  (** Return edge is between the return statement, which may optionally contain
+    * a return value, and the function. The result of the call is then
+    * transferred to the function node! *)
+  | Test of CilType.Exp.t * bool
+  (** The true-branch or false-branch of a conditional exp *)
+  | ASM of string list * asm_out * asm_in
+  (** Inline assembly statements, and the annotations for output and input
+    * variables. *)
+  | VDecl of CilType.Varinfo.t
+  (** VDecl edge for the variable in varinfo. Whether such an edge is there for all
+    * local variables or only when it is not possible to pull the declaration up, is
+    * determined by alwaysGenerateVarDecl in cabs2cil.ml in CIL. One case in which a VDecl
+    * is always there is for VLA. If there is a VDecl edge, it is where the declaration originally
+    * appeared *)
+  | Skip
+  (** This is here for historical reasons. I never use Skip edges! *)
+  | SelfLoop
+  (** This for interrupt edges.! *)
+[@@deriving to_yojson]
+
+
+let pretty () = function
+  | Assign (lv,rv) -> dprintf "Assign '%a = %a' " d_lval lv d_exp rv
+  | Proc (None  ,f,ars) -> dprintf "Proc '%a(%a)'" d_exp f (d_list ", " d_exp) ars
+  | Proc (Some r,f,ars) -> dprintf "Proc '%a = %a(%a)'" d_lval r d_exp f (d_list ", " d_exp) ars
+  | Entry f -> dprintf "Entry %s" f.svar.vname
+  | Ret (None,fd) -> dprintf "Ret (None, %s)" fd.svar.vname
+  | Ret (Some r,fd) -> dprintf "Ret (Some %a, %s)" d_exp r fd.svar.vname
+  | Test (p,b) -> dprintf "Test (%a,%b)" d_exp p b
+  | ASM _ -> text "ASM ..."
+  | Skip -> text "Skip"
+  | VDecl v -> dprintf "VDecl '%a %s;'" d_type v.vtype v.vname
+  | SelfLoop -> text "SelfLoop"
+
+(* TODO: unused *)
+let pretty_kind () = function
+  | Assign (lv,rv) -> dprintf "Assign"
+  | Proc (_  ,f,ars) -> dprintf "Proc"
+  | Entry f -> dprintf "Entry %s" f.svar.vname
+  | Ret (r,fd) -> dprintf "Ret"
+  | Test (p,b) -> dprintf "Test"
+  | ASM _ -> text "ASM"
+  | Skip -> text "Skip"
+  | VDecl _ -> text "VDecl"
+  | SelfLoop -> text "SelfLoop"
```

Comment:
If it unused should it be removed?

Reply:
I might do that once I figure out the similar situation in `Node`, where at least four different output formats (some with multiple copy-pasted implementations) exist:

https://github.com/goblint/analyzer/blob/ef44a3c988cee5a930f984d180b6f96db773e57a/src/framework/node.ml#L16-L26
https://github.com/goblint/analyzer/blob/ef44a3c988cee5a930f984d180b6f96db773e57a/src/framework/analyses.ml#L47-L51
https://github.com/goblint/analyzer/blob/ef44a3c988cee5a930f984d180b6f96db773e57a/src/framework/cfgTools.ml#L390-L394

I could move all the variants together into `Node` there to avoid identical duplicate implementations but it still seems excessive. Seems like we should have one consistent format, but I don't know which one it should be. Maybe the last CFG node naming one is the best: just `sid` for `Statement`, `f()` for `FunctionEntry f` and `return of f()` for `Function f`.
Others seem a bit more confusing, like `call of f()` meaning the return node.

Also other `Statement` ones include the statement itself, which might sometimes be useful in terminal output instead of just a node ID that needs to be looked up in the HTML output. But statement printing is a bit weird as well: for `Instr`s it's just one line for sure, but for `If` that's actually the whole if statement with both branches printed by `Cil`, which is multiline and super verbose.
And tbh, I've always found statements on nodes extremely confusing because what the hell do they even mean? Statements correspond to the edges of the CFG, not nodes between them. The node itself isn't executing any statement. So it's unclear whether the statement printed for a node corresponds to an edge before or after that node.
I believe the way our CFG is constructed, the statement in the node is the same as the edge going out of it. But then during analysis, the transfer function has `ctx.node` which corresponds to the node being evaluated while `ctx.prev_node` is the one `ctx.local` corresponds to. So actually the statement in `prev_node` is the one being evaluated and the statement in `node` has nothing to do with the current transfer function.

Reply:
Since there's no clear best way to output nodes, I just moved all the possible variants into `Node` after all. But at least gave them sensible names and documentation.

The unused `Edge` thing is also removed now.